src/cpp/rtps/reader/StatelessReader.cpp

```diff
@@ -518,9 +410,23 @@ bool StatelessReader::processDataMsg(
 
         // Ask payload pool to copy the payload
         IPayloadPool* payload_owner = change->payload_owner();
-        ReaderPool* datasharing_pool = dynamic_cast<ReaderPool*>(payload_owner);
-        if (datasharing_pool)
+
+        if (is_datasharing_compatible_ && datasharing_listener_->writer_is_matched(change->writerGUID))
         {
+            //We may receive the change from the listener (with owner a ReaderPool) or intraprocess (with owner a WriterPool)
+            ReaderPool* datasharing_pool = dynamic_cast<ReaderPool*>(payload_owner);
+            if (!datasharing_pool)
+            {
+                datasharing_pool = datasharing_listener_->get_pool_for_writer(change->writerGUID).get();
+            }
+            if (!datasharing_pool)
+            {
+                logWarning(RTPS_MSG_IN, IDSTRING "Problem copying DataSharing CacheChange from writer "
+                        << change->writerGUID);
+                change_pool_->release_cache(change_to_add);
```

Comment:
I don't think we should release the change in this case.

Reply:
We reserved the change in line 403, we must release it before exiting.

Are you suggesting that we process the change as if not coming from a remote (non-datasharing) writer? I think is best not to. At this point, we are on an inconsistent state: The writer is matched as datasharing, but the change is not form a datasharing pool, we should directly fail the method.

Reply:
I misread `change_pool_` as `payload_pool_`